### ISSUE TITLE
Unnecessary CSS in Jupyter Notebook

### ISSUE THREAD

**Issue by kitchensjn** [kitchensjn/tskit_arg_visualizer#39]
Each time the visualizer is run in the Jupyter Notebook reopens and adds the CSS stylings again. This leads to a cluttering of the web inspector with unnecessary (redundant) CSS. Look into ways of minimizing this - maybe there's a way to check if it's already been added to the browser before adding again.

**Comment by kitchensjn** [comment]
This can be solved by loading the CSS separately when you import rather than every time you draw.

```
import tskit_arg_visualizer
tskit_arg_visualizer.load_styles()
```

Unfortunately, you can't set this function to run automatically when you import the module because there are instances where the module has been imported previously but for a different notebook. So far I have been unable to find a way to check whether the styles have been loaded in the browser.

**Comment by kitchensjn** [comment]
Since I've confused myself with this - this is not yet implemented in any version or branch. This is purely a theoretical example of what we could do to reduce the amount of unnecessary CSS in the notebooks.

**Comment by hyanwong** [comment]
It's not just the CSS, but also the generalised javascript functions that get redefined when each `draw` or `draw_nodes` command is called, I think? It would be nice if these too could be defined just once in each notebook.

**Comment by hyanwong** [comment]
What if we set an attribute on the D3ARG class the first time a viz is drawn, and only include the styles if the attribute doesn't exist?

```python
def draw_D3(arg_json, force_notebook=False):
    ...
    css_file = os.path.dirname(__file__) + "/visualizer.css"
    if force_notebook or running_in_notebook():
        if not hasattr(D3ARG, "loaded_styles"):  # Check if the styles have already been loaded
            with open(css_file, r) as css:
                display(HTML("<style>"+css.read()+"</style>"))
            D3ARG.loaded_styles = True  # Add a static flag to the whole class
        display(HTML(html))

```

**Comment by kitchensjn** [comment]
That could work. I also have thought this could tie into #117 where rather than having a giant template, you only have a template with the function calls. For instance, instead of the current implementation

```
function draw_force_diagram() {
    ...
    $width
    ...
    $height
}
```

you can first predefine all of your functions then call the with various template calls

```
function draw_force_diagram(width, height) {
    ...
}

draw_force_diagram($width, $height)
```

All of the functions can be loaded separately like a package so they are only defined once, and then the template is just calls to those functions. This seems like a much cleaner way of handling things compared to the current implementation and should scale up more easily as features are continually added as write now it's lacking a lot of modularity. Relatively big undertaking to make this change, but I think had I known more JS when I started the project, I would have built a cleaner foundation like this. Thoughts?

**Comment by hyanwong** [comment]
Yes, that sounds doable, and at least for a first pass, not that much work, actually. Do you want me to have a go? It could also fix #143

**Comment by hyanwong** [comment]
I had a quick go. I think you actually want to define the template variables in the highest-level function:

```javascript
function main_visualizer(
    d3,
    divnum,
    graph,
    width,
    height,
    y_axis,
    edge_styles,
    condense_mutations,
    include_mutation_labels,
    tree_highlighting,
    title,
    rotate_tip_labels,
    plot_type,
    source, 
) {
    /*! @source http://purl.eligrey.com/github/FileSaver.js/blob/master/FileSaver.js */
    
    function download (url, name, opts) {
        ...
```

Then you can template the following snippet, which is the only JS you need to insert repeatedly for each plot. This template could be hard-coded into __init__.py, or reside in a different JS file.

```javascript
ensureRequire()
    .then(require => {
        require.config({ paths: {d3: 'https://d3js.org/d3.v7.min'}});
        require(["d3"], function(d3) {
            main_visualizer(d3, $divnum, $data, $width, $height, $y_axis, $edges, $condense_mutations, $include_mutation_labels, $tree_highlighting, $title, $rotate_tip_labels, $plot_type, $source);
        })
    })
    .catch(err => console.error('Failed to load require.js:', err));
```

**Comment by hyanwong** [comment]
The function reworking is now implemented, but we haven't yet implemented a method to reduce code duplication in different cells.
